oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,402 @@
+  aren't strictly needed for any of the regular contexts (``private.in``).
+
+.. _Sphinx: http://www.sphinx-doc.org/
+
+Declare Direct Dependencies
+---------------------------
+
+As indicated above, some of the usage contexts have a standard filename used in
+the ``requirements`` directory of an Open edX repository to list dependencies.
+Others will have an appropriate filename custom to that repository's unique
+context.  Each of these is a ``pip``-compatible `requirements file`_ listing
+the direct dependencies needed for that context.  Beyond complying with the
+file format, there are a few guidelines each of these files should follow:
+
+* The file should start with a brief comment explaining the context in which
+  these dependencies are needed.  Examples can be found in the
+  `cookiecutter-django-app`_ repository.
+* Each listed dependency should have a brief end-of-line comment explaining
+  its primary purpose(s) in this context.  These comments typically start at
+  the 37th character, which allows enough room for most package name plus
+  constraint specifications while keeping the comments visually aligned.
+* Version constraints should only be used to exclude dependency versions which
+  are known (or strongly suspected) to not work in this context.
+* Indirect dependencies (used by dependencies but not directly by the code in
+  the repository itself) should not be listed unless a constraint is needed to
+  enforce a compatible version; these are automatically detected and captured
+  elsewhere as described below.
+* `Environment markers`_ should be used as necessary to indicate dependencies
+  which should only be installed on specific operating systems, Python
+  versions, etc.
+* Avoid direct links to packages in local directories, GitHub, or other version
+  control systems if at all possible; all dependencies should be installed
+  from `PyPI`_.  If you think you're in one of the rare circumstances where
+  installing a package from a URL is appropriate, see the notes below on
+  `Installing Dependencies from URLs`_
+* If the dependencies in one context are a superset of those in another one,
+  do not repeat the dependencies.  Instead, explicitly include the file
+  produced by ``make upgrade`` for the smaller set of dependencies in the
+  requirements file for the larger set of dependencies. For example,
+  ``test.in`` often includes a line like the following to ensure that the same
+  versions of packages used in production for a service will also be used when
+  testing it:
+
+.. code-block:: python
+
+  -r base.txt                         # Core dependencies of the service being tested
+
+If the repository contains a ``setup.py`` file defining a Python package, the
+base dependencies also need to be specified there.  These can be derived from
+``requirements/base.in`` with a simple Python function declared in
+``setup.py`` itself.  An example can be found in the
+`setup.py file for edx-completion`_.
```

Comment:
Would it be better to inline the function definition in this document? That would give a more canonical best-version, it seems like.

Reply:
Fair enough.  The first draft of the OEP didn't particularly recommend this approach, but earlier review discussions concluded that it was best not to explicitly list the core dependencies in two places, so having a canonical implementation of loading them makes sense.  The existing code is about 30 lines; a little long, but probably short enough to include.